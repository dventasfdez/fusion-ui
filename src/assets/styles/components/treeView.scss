.fusion-ui {
  .tree-container,
  &.tree-container {
    @extend .list-icon;
    width: 100%;
  }

  li.tree-element {
    cursor: pointer;
    @include font(medium);
    border-left: calc($unit / 4) solid transparent;
    height: auto;
    width: 100%;
    display: flex;
    align-items: center;
    &::before {
      content: none;
    }

    .tree-element-icon-right {
      margin-left: auto;
      font-size: calc(2 * $unit);
    }

    &:hover {
      color: var(--text-interactive-link-on-light-hover);
      border-left: calc($unit / 4) solid transparent;
    }

    &:active {
      border-left: calc($unit / 4) solid transparent;
      color: var(--text-interactive-link-on-light-pressed);
    }

    &:focus-visible {
      outline: calc($unit / 2) solid var(--borders-dividers-interactive-focused);
    }

    &[data-disabled="true"] {
      pointer-events: none;
      border-left: calc($unit / 4) solid transparent;
      color: var(--text-03);
    }
  }
}
