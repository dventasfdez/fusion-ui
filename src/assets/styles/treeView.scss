.tag-ds {
  .tree-container,
  &.tree-container {
    width: 100%;
    .tree-element,
    &.tree-element {
      margin-top: 0;
      margin-bottom: 0;
      margin-bottom: $unit;
      @include font(medium);
      text-transform: none;
      border: 0;
      border-left: calc($unit / 4) solid transparent;
      background-color: transparent;
      border-radius: 0;
      padding: 0;
      padding-left: calc(3 * $unit);
      height: auto;
      width: 100%;
      &:has(.material-icons:not(.left, .right)) {
        padding: 0;
        padding-left: calc(3 * $unit);
        height: auto;
      }

      .tree-li {
        padding: 0;
        display: flex;
        align-items: center;
        padding-right: calc(2 * $unit);
      }

      & .list-icon {
        margin-top: 0;
        margin-bottom: 0;
        padding-top: 0;
      }

      .tree-element-icon-right {
        margin-left: auto;
        font-size: calc(2 * $unit);
      }

      & > li {
        color: var(--text-01);
        text-align: left;
        width: 100%;
      }

      &:hover {
        & > li {
          color: var(--text-interactive-link-on-light-hover);
        }
        background-color: transparent;
        border: 0;
        border-left: calc($unit / 4) solid transparent;
      }

      &:active {
        border: 0;
        border-left: calc($unit / 4) solid transparent;
        & > li {
          color: var(--text-interactive-link-on-light-pressed);
        }
      }

      //border-left: calc($unit / 4) solid var(--surface-accent);
      &:focus-visible {
        outline: calc($unit / 2) solid var(--borders-dividers-interactive-focused);
      }

      &:disabled {
        background-color: transparent;
        border: 0;
        border-left: calc($unit / 4) solid transparent;
        & > li {
          color: var(--text-03);
        }
      }
    }

    .tree-element-item {
      padding-left: $unit;
    }
  }
}
