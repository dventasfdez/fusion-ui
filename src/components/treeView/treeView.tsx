import React from 'react';
import TreeViewElement from './treeViewElement';
export {default as TreeViewElement} from './treeViewElement';

interface ITree {
  /**
   * Alternative className for Tree
   */
  className?: string;
  [other: string]: any;
}

const Tree: React.FC<ITree> = ({children, className, ...rest}) => {
  const renderable = React.Children.toArray(children).every((child: any) => child.type === TreeViewElement);
  return renderable ? (
    <div className={`tree-container ${className || ''}`} {...rest}>
      {children}
    </div>
  ) : null;
};

export default Tree;
