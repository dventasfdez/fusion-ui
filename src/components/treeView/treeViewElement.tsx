import React, {useState} from 'react';

export interface ITreeElement {
  /**
   * Name of the Tree Element
   */
  name: string;
  /**
   * For disabling the Tree Element
   */
  disabled?: boolean;
  /**
   * click event on developers needs
   */
  onClick?: (e?: React.MouseEvent) => void;
  /**
   * Alternative className on Tree.Element
   */
  className?: string;
  /**
   * prop for displaying an icon if needed
   */
  icon?: string;
  /**
   * By default left sided arrow, can be passed an oposite side (both will show up if they have children)
   */
  arrow?: 'left' | 'right' | 'none';
  [others: string]: any;
}

const TreeViewElement: React.FC<ITreeElement> = ({
  children,
  name,
  onClick,
  disabled,
  className,
  arrow = 'left',
  icon,
  ...rest
}) => {
  const [showChildren, setShowChildren] = useState<boolean>(false);

  const handleClickElement = (e: React.MouseEvent) => {
    setShowChildren(!showChildren);
    if (typeof onClick === 'function') onClick(e);
  };

  const renderIcon = () => {
    if (icon) return icon;
    if (!icon && arrow === 'left') return 'arrow_right';
    return '';
  };

  return (
    <>
      <button
        type="button"
        disabled={disabled}
        className={`list-icon tree-element ${className || ''}`}
        onClick={handleClickElement}
        {...rest}
      >
        <li className="item-condensed tree-li">
          {children || icon ? (
            <span className="material-icons icon-order">{renderIcon()}</span>
          ) : (
            <span className="icon-order" />
          )}
          {name}
          {arrow === 'right' &&
            children &&
            (showChildren ? (
              <span className="material-icons tree-element-icon-right">keyboard_arrow_up</span>
            ) : (
              <span className="material-icons tree-element-icon-right">keyboard_arrow_down</span>
            ))}
        </li>
      </button>
      {children &&
        showChildren &&
        React.Children.toArray(children).map((_child, _idx) => (
          <div key={_idx} className="tree-element-item">
            {_child}
          </div>
        ))}
    </>
  );
};

export default TreeViewElement;
