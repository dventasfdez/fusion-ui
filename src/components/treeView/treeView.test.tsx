import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import {Tree, TreeViewElement} from './treeView';
import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const TreeViewTest = (args: any) => {
  return (
    <div className="tag-ds">
      <Tree>
        <TreeViewElement data-testid="tree-element-1" name="Item Tree"></TreeViewElement>
        <TreeViewElement name="Item Tree"></TreeViewElement>
        {args.icon && <TreeViewElement data-testid="tree-element-5" name="Item Tree" icon={args.icon}></TreeViewElement>}
        <TreeViewElement data-testid="tree-element-2" name="Item Tree">
          <TreeViewElement data-testid="tree-element-3" name="Sub item">
            <TreeViewElement data-testid="tree-element-4" onClick={args.onClick} name="Sub item 2" />
          </TreeViewElement>
        </TreeViewElement>
      </Tree>
    </div>
  );
};

const TreeViewTestNonRenderable = (args: any) => {
  return (
    <div className="tag-ds">
      <Tree>
        <li data-testid="tree-element-1">item tree</li>
      </Tree>
    </div>
  );
};

test('render treeView and match snapshot', () => {
  const component = renderer.create(<TreeViewTest />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('on click function Tree Element', () => {
  const fn = jest.fn();
  const {getByTestId} = render(<TreeViewTest onClick={fn} />);
  userEvent.click(getByTestId('tree-element-2'));
  userEvent.click(getByTestId('tree-element-3'));
  userEvent.click(getByTestId('tree-element-4'));
  expect(fn).toBeCalledTimes(1);
});

test('non renderable Tree Element', () => {
  const {queryByText} = render(<TreeViewTestNonRenderable />);
  const element = queryByText(/item tree/i);
  expect(element).toBeNull();
});

test('Tree Element with icon', () => {
  const {getByTestId} = render(<TreeViewTest icon="home" />);
  const element = getByTestId('tree-element-1');
  expect(element).toHaveClass('tree-element');
});
