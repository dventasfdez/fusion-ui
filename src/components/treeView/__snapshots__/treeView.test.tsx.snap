// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render treeView and match snapshot 1`] = `
<div
  className="tag-ds"
>
  <div
    className="tree-container "
  >
    <button
      className="list-icon tree-element "
      data-testid="tree-element-1"
      onClick={[Function]}
      type="button"
    >
      <li
        className="item-condensed tree-li"
      >
        <span
          className="icon-order"
        />
        Item Tree
      </li>
    </button>
    <button
      className="list-icon tree-element "
      onClick={[Function]}
      type="button"
    >
      <li
        className="item-condensed tree-li"
      >
        <span
          className="icon-order"
        />
        Item Tree
      </li>
    </button>
    <button
      className="list-icon tree-element "
      data-testid="tree-element-2"
      onClick={[Function]}
      type="button"
    >
      <li
        className="item-condensed tree-li"
      >
        <span
          className="material-icons icon-order"
        >
          arrow_right
        </span>
        Item Tree
      </li>
    </button>
  </div>
</div>
`;
